File: web/src/lib/api/devices.ts
import { logger } from '../logger';
import { getApiUrl, TAG, AuthError, getToken, isAuthFailure, withAuthRetry, fetchWithDeviceAuth } from './client';

export type DeviceDto = {
  /** 1–999 per-user display number from server; absent for LAN-only rows. */
  displayCode?: number | null;
  deviceId: string;
  name: string;
  platform?: string | null;
  lanHttpUrl?: string | null;
  lastSeen?: number | null;
  presenceStatus?: 'online' | 'offline' | string | null;
  presenceUpdatedAt?: number | null;
};

export async function registerDevice(deviceId: string, name: string, opts?: { lanHttpUrl?: string; platform?: string; sessionId?: string }): Promise<DeviceDto> {
  logger.info(TAG, 'registerDevice deviceId=', deviceId);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const res = await fetch(`${getApiUrl()}/api/devices`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify({
        deviceId,
        name,
        lanHttpUrl: opts?.lanHttpUrl || undefined,
        platform: opts?.platform || undefined,
        sessionId: opts?.sessionId || undefined,
      }),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'registerDevice failed', res.status);
      throw new Error('Failed to register device');
    }
    const dto = await res.json() as DeviceDto;
    logger.info(TAG, 'registerDevice success deviceId=', dto.deviceId);
    return dto;
  });
}

export async function updateDevicePresence(
  deviceId: string,
  body: { sessionId: string; status: 'online' | 'offline'; platform?: string },
  opts?: { keepalive?: boolean },
): Promise<DeviceDto> {
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const res = await fetch(`${getApiUrl()}/api/devices/${encodeURIComponent(deviceId)}/presence`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify(body),
      keepalive: opts?.keepalive,
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'updateDevicePresence failed', res.status);
      throw new Error('Failed to update device presence');
    }
    return res.json();
  });
}

export async function updateDevice(deviceId: string, data: { name?: string; lanHttpUrl?: string }): Promise<DeviceDto> {
  logger.info(TAG, 'updateDevice deviceId=', deviceId);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const res = await fetch(`${getApiUrl()}/api/devices/${encodeURIComponent(deviceId)}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify(data),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'updateDevice failed', res.status);
      throw new Error('Failed to update device');
    }
    return res.json();
  });
}

export async function deleteDevice(deviceId: string): Promise<void> {
  logger.info(TAG, 'deleteDevice deviceId=', deviceId);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const res = await fetch(`${getApiUrl()}/api/devices/${encodeURIComponent(deviceId)}`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'deleteDevice failed', res.status);
      throw new Error('Failed to delete device');
    }
    logger.info(TAG, 'deleteDevice success deviceId=', deviceId);
  });
}

export async function listDevices(): Promise<DeviceDto[]> {
  logger.info(TAG, 'listDevices');
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const res = await fetch(`${getApiUrl()}/api/devices`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'listDevices failed', res.status);
      throw new Error('Failed to list devices');
    }
    const list = await res.json() as DeviceDto[];
    logger.info(TAG, 'listDevices success count=', list.length);
    return list;
  });
}

export async function listPairedDevices(): Promise<DeviceDto[]> {
  const res = await fetchWithDeviceAuth(`${getApiUrl()}/api/devices/paired`);
  if (!res.ok) throw new Error('Failed to list paired devices');
  return res.json();
}

export async function unpairDevice(peer: string): Promise<void> {
  const res=await fetchWithDeviceAuth(`${getApiUrl()}/api/devices/paired/${encodeURIComponent(peer)}`,{method:'DELETE'});
  if(!res.ok) throw new Error('Failed to disconnect device');
}

File: app/lib/api/devices.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../logger.dart';
import 'client.dart';

class DeviceDto {
  /// Server-assigned 1–999 display number per user; null for LAN-only rows.
  final int? displayCode;
  final String deviceId;
  final String name;
  final String? platform;
  final String? lanHttpUrl;
  final String? presenceStatus;
  final int? presenceUpdatedAt;

  /// Epoch millis; null if never seen. Used to show online (e.g. within last 2 min).
  final int? lastSeen;
  DeviceDto({
    this.displayCode,
    required this.deviceId,
    required this.name,
    this.platform,
    this.lanHttpUrl,
    this.presenceStatus,
    this.presenceUpdatedAt,
    this.lastSeen,
  });
  factory DeviceDto.fromJson(Map<String, dynamic> j) => DeviceDto(
    displayCode: _readOptionalInt(j['displayCode'] ?? j['display_code']),
    deviceId: j['deviceId'] as String,
    name: j['name'] as String,
    platform: j['platform'] as String?,
    lanHttpUrl: j['lanHttpUrl'] as String?,
    presenceStatus: j['presenceStatus'] as String?,
    presenceUpdatedAt: j['presenceUpdatedAt'] is int
        ? j['presenceUpdatedAt'] as int
        : null,
    lastSeen: j['lastSeen'] is int ? j['lastSeen'] as int : null,
  );

  static int? _readOptionalInt(dynamic v) {
    if (v == null) return null;
    if (v is int) return v;
    if (v is num) return v.toInt();
    return int.tryParse(v.toString());
  }

  bool get isWeb => platform == 'web';
}

Future<List<DeviceDto>> listDevices() async {
  logApi.info('listDevices');
  return withAuthRetry(() async {
    final r = await http.get(
      Uri.parse('$apiBaseUrl/api/devices'),
      headers: apiHeaders,
    );
    checkAuthResponse(r, fallback: '获取设备列表失败');
    final list = (jsonDecode(r.body) as List)
        .map((e) => DeviceDto.fromJson(e as Map<String, dynamic>))
        .toList();
    logApi.info('listDevices success count=${list.length}');
    return list;
  });
}

Future<DeviceDto> registerDevice(
  String deviceId,
  String name, {
  String? lanHttpUrl,
  String? platform,
  String? sessionId,
}) async {
  logApi.info('registerDevice deviceId=$deviceId');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/devices'),
      headers: apiHeaders,
      body: jsonEncode({
        'deviceId': deviceId,
        'name': name,
        if (lanHttpUrl != null) 'lanHttpUrl': lanHttpUrl,
        if (platform != null) 'platform': platform,
        if (sessionId != null) 'sessionId': sessionId,
      }),
    );
    checkAuthResponse(r, fallback: '设备注册失败');
    final dto = DeviceDto.fromJson(jsonDecode(r.body) as Map<String, dynamic>);
    logApi.info('registerDevice success deviceId=${dto.deviceId}');
    return dto;
  });
}

Future<DeviceDto> updateDevicePresence(
  String deviceId, {
  required String sessionId,
  required String status,
  String? platform,
}) async {
  logApi.info('updateDevicePresence deviceId=$deviceId status=$status');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse(
        '$apiBaseUrl/api/devices/${Uri.encodeComponent(deviceId)}/presence',
      ),
      headers: apiHeaders,
      body: jsonEncode({
        'sessionId': sessionId,
        'status': status,
        if (platform != null) 'platform': platform,
      }),
    );
    checkAuthResponse(r, fallback: '更新设备在线状态失败');
    return DeviceDto.fromJson(jsonDecode(r.body) as Map<String, dynamic>);
  });
}

Future<void> deleteDevice(String deviceId) async {
  logApi.info('deleteDevice deviceId=$deviceId');
  return withAuthRetry(() async {
    final r = await http.delete(
      Uri.parse('$apiBaseUrl/api/devices/${Uri.encodeComponent(deviceId)}'),
      headers: apiHeaders,
    );
    checkAuthResponse(r, fallback: '删除设备失败');
    logApi.info('deleteDevice success deviceId=$deviceId');
  });
}

Future<DeviceDto> updateDevice(
  String deviceId, {
  String? name,
  String? lanHttpUrl,
}) async {
  logApi.info('updateDevice deviceId=$deviceId');
  return withAuthRetry(() async {
    final r = await http.patch(
      Uri.parse('$apiBaseUrl/api/devices/${Uri.encodeComponent(deviceId)}'),
      headers: apiHeaders,
      body: jsonEncode({
        if (name != null) 'name': name,
        if (lanHttpUrl != null) 'lanHttpUrl': lanHttpUrl,
      }),
    );
    checkAuthResponse(r, fallback: '更新设备失败');
    return DeviceDto.fromJson(jsonDecode(r.body) as Map<String, dynamic>);
  });
}

Future<List<DeviceDto>> listPairedDevices() async {
  if (!hasDeviceAccessToken) return [];
  final r = await withDeviceAuthRetry(
    () => http.get(
      Uri.parse('$apiBaseUrl/api/devices/paired'),
      headers: deviceApiHeaders,
    ),
  );
  if (r.statusCode != 200) throw Exception('Failed to list paired devices');
  return (jsonDecode(r.body) as List)
      .map((e) => DeviceDto.fromJson(e as Map<String, dynamic>))
      .toList();
}

Future<void> unpairDevice(String peer) async {
  final r = await withDeviceAuthRetry(
    () => http.delete(
      Uri.parse('$apiBaseUrl/api/devices/paired/${Uri.encodeComponent(peer)}'),
      headers: deviceApiHeaders,
    ),
  );
  if (r.statusCode != 204) throw Exception('Failed to disconnect device');
}
